lib/charms/operator_libs_linux/v1/systemd.py

```diff
@@ -66,127 +70,65 @@
-        quiet: passes the --quiet flag to the shell invocation.
+        *args: Arguments to pass to systemctl.
     """
-    cmd = ["systemctl", sub_cmd]
-
-    if service_name is not None:
-        cmd.append(service_name)
-    if now is not None:
-        cmd.append("--now")
-    if quiet is not None:
-        cmd.append("--quiet")
-    if sub_cmd != "is-active":
-        logger.debug("Attempting to {} '{}' with command {}.".format(cmd, service_name, cmd))
-    else:
-        logger.debug("Checking if '{}' is active".format(service_name))
-
-    proc = subprocess.Popen(cmd, **_popen_kwargs())
-    last_line = ""
-    for line in iter(proc.stdout.readline, ""):
-        last_line = line
-        logger.debug(line)
-
-    proc.wait()
-
-    if proc.returncode < 1:
-        return True
-
-    # If we are just checking whether a service is running, return True/False, rather
-    # than raising an error.
-    if sub_cmd == "is-active" and proc.returncode == 3:  # Code returned when service not active.
+    cmd = ["systemctl", *args]
+    logger.debug(f"Executing command: {cmd}")
+    proc = subprocess.run(
+        [*cmd],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        text=True,
+        bufsize=1,
+        encoding="utf-8"
+    )
+    logger.debug(proc.stdout)
+    if proc.returncode == 1 and "is-failed" in cmd:
         return False
-
-    if sub_cmd == "is-failed":
+    elif proc.returncode == 3 and "is-active" in cmd:
         return False
-
-    raise SystemdError(
-        "Could not {}{}: systemd output: {}".format(
-            sub_cmd, " {}".format(service_name) if service_name else "", last_line
-        )
-    )
-
-
-def service_running(service_name: str) -> bool:
-    """Determine whether a system service is running.
-
-    Args:
-        service_name: the name of the service to check
-    """
-    return _systemctl("is-active", service_name, quiet=True)
-
-
-def service_failed(service_name: str) -> bool:
-    """Determine whether a system service has failed.
-
-    Args:
-        service_name: the name of the service to check
-    """
-    return _systemctl("is-failed", service_name, quiet=True)
-
-
-def service_start(service_name: str) -> bool:
-    """Start a system service.
-
-    Args:
-        service_name: the name of the service to start
-    """
-    return _systemctl("start", service_name)
-
-
-def service_stop(service_name: str) -> bool:
-    """Stop a system service.
-
-    Args:
-        service_name: the name of the service to stop
-    """
-    return _systemctl("stop", service_name)
-
-
-def service_restart(service_name: str) -> bool:
-    """Restart a system service.
-
-    Args:
-        service_name: the name of the service to restart
-    """
-    return _systemctl("restart", service_name)
+    elif proc.returncode == 0:
+        return True
+    else:
+        raise SystemdError(f"Command {cmd} failed. systemctl output:\n{proc.stdout}")
```

Comment:
Hey Ben! So I mostly had the evaluation of the the return codes like this in `_systemctl(...)` because that's how it was in the earlier version of the `_systemctl(...)` :sweat_smile:  

I agree that it feels weird that we are following an if-else chain within `_systemctl` and modifying the return value based on specific use-cases. I like the idea of using a `check=False` kwarg to direct `_systemctl(...)` to raise an error or not. I do believe that _systemctl should just return the return code since it should handle logging and raising the SystemdError itself.

Reply:
`service_running(...)` and `service_failed(...)` now evaluate the return codes themselves to determine whether `True`/`False` should be returned rather than relying on _systemctl to filter the `subprocess.CompletedProcess` returned by `subprocess.run`. This change makes the implementation of `_systemctl(...)` much simpler :sweat_smile: 

Change made in commit https://github.com/canonical/operator-libs-linux/pull/106/commits/af8c1fb3b34fbfc19b74b20d4210dd45c7eb07db